### ISSUE TITLE
`alloc_slice_dst_in` may be unsound

### ISSUE THREAD

**Issue by danielhenrymantilla** [CAD97/pointer-utils#23]
Given the safety requirements of [`GlobalAlloc::alloc()`](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.alloc) (and thus of `alloc::alloc`),
the following pattern looks unsound:

https://github.com/CAD97/pointer-utils/blob/a0d6f5e2e6998465a57751e3eeec8955682a29e7/crates/slice-dst/src/lib.rs#L140-L146

What if `container = |_| Layout::new::<()>()`?

**Comment by CAD97** [comment]
Yikes, good catch! Yeah, this is currently unsound per the current formulation 😬. In fact, `alloc_slice_dst` currently relies on `<_ as SliceDst>::layout_for` not returning an empty layout (which is not specified as a requirement).

Here's a test that fails under miri (after adding `SliceDst for [_]`, which I don't know why I missed):

```rust
#[test]
fn actual_zst() {
    unsafe {
        let slice: Box<[MaybeUninit<()>]> = Box::new_slice_dst(10, drop);
    }
}
```

<details><summary>I did try to test the ZST case, I just tested the wrong one</summary>

```rust
#[test]
fn zst() {
    let slice: Vec<()> = vec![(); 16];
    let slice: Box<SliceWithHeader<(), ()>> = SliceWithHeader::new((), slice);
    let slice = slice.clone();
}
```
</details>

Thankfully, I think this is fixable with only changes to the implementation of `alloc_slice_dst_in`, and a documentation note of the outcome of allocating for a ZST. The patch I believe fixes this:

```diff
diff --git a/crates/slice-dst/src/lib.rs b/crates/slice-dst/src/lib.rs
index 62d233c..1d3a6bc 100644
--- a/crates/slice-dst/src/lib.rs
+++ b/crates/slice-dst/src/lib.rs
@@ -139,6 +139,9 @@ unsafe impl<T> SliceDst for [T] {
 ///
 /// The returned pointer is owned and completely uninitialized;
 /// you are required to initialize it correctly.
+///
+/// If the type to be allocated has zero size,
+/// then an arbitrary aligned dangling nonnull pointer is returned.
 pub fn alloc_slice_dst<S: ?Sized + SliceDst>(len: usize) -> ptr::NonNull<S> {
     alloc_slice_dst_in(|it| it, len)
 }
@@ -147,14 +150,26 @@ pub fn alloc_slice_dst<S: ?Sized + SliceDst>(len: usize) -> ptr::NonNull<S> {
 ///
 /// The returned pointer is owned and completely uninitialized;
 /// you are required to initialize it correctly.
+///
+/// Note that while this function returns a `ptr::NonNull<S>`,
+/// the pointer is to the allocation as specified by `container(S::layout(len))`,
+/// so if you want/need a pointer to `S`, you will need to offset it.
+///
+/// If the layout` to be allocated has zero size,
+/// then an arbitrary aligned dangling nonnull pointer is returned.
 pub fn alloc_slice_dst_in<S: ?Sized + SliceDst, F>(container: F, len: usize) -> ptr::NonNull<S>
 where
     F: FnOnce(Layout) -> Layout,
 {
     let layout = container(S::layout_for(len));
     unsafe {
-        let ptr = ptr::NonNull::new(alloc(layout) as *mut ())
-            .unwrap_or_else(|| handle_alloc_error(layout));
+        let ptr = if layout.size() == 0 {
+            // Do not allocate in the ZST case! CAD97/pointer-utils#23
+            ptr::NonNull::new(layout.align() as *mut ())
+        } else {
+            ptr::NonNull::new(alloc(layout) as *mut ())
+        }
+        .unwrap_or_else(|| handle_alloc_error(layout));
         let ptr = ptr::NonNull::new_unchecked(slice::from_raw_parts_mut::<()>(ptr.as_ptr(), len));
         S::retype(ptr)
     }
```

**Comment by danielhenrymantilla** [comment]
Good job!

**Comment by CAD97** [comment]
[slice-dst:1.1.0](https://crates.io/crates/slice-dst/1.1.0) is published and v1.0.0 is yanked 🎉
